flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java

```diff
@@ -0,0 +1,209 @@
+
+					IOUtils.copyBytes(new ByteArrayInputStream(testFileContent.getBytes(StandardCharsets.UTF_8)), zis, false);
+				}
+				return zipArchive;
+			} else {
+				throw new IllegalArgumentException("This service contains only entry for " + permanentBlobKey);
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+
+		}
+	};
+
+	@Before
+	public void setup() throws Exception {
+		fileCache = new FileCache(new String[]{temporaryFolder.newFolder().getAbsolutePath()}, blobService);
+	}
+
+	@After
+	public void shutdown() {
+		fileCache.shutdown();
+	}
+
+	@Test
+	public void testDirectoryDownloadedFromBlob() throws Exception {
+		JobID jobID = new JobID();
+		ExecutionAttemptID attemptID = new ExecutionAttemptID();
+
+		final String fileName = "test_file";
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey),
+			true);
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID, attemptID);
+
+		final Path dstPath = copyResult.get();
+		final FileSystem fs = dstPath.getFileSystem();
+		final FileStatus fileStatus = fs.getFileStatus(dstPath);
+		assertTrue(fileStatus.isDir());
+
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fs.exists(cacheFile));
+		final String actualContent = FileUtils.readFileUtf8(new File(cacheFile.getPath()));
+		assertEquals(testFileContent, actualContent);
+	}
+
+	@Test
+	public void testDirectoryCleanUp() throws Exception {
+		fileCache.shutdown();
+		DeleteCapturingDirectScheduledExecutorService scheduledExecutorService =
+			new DeleteCapturingDirectScheduledExecutorService();
+
+		final int cleanupInterval = 1000;
+		fileCache = new FileCache(new String[]{temporaryFolder.newFolder().getAbsolutePath()}, blobService,
+			scheduledExecutorService, cleanupInterval);
+
+		JobID jobID = new JobID();
+		ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		ExecutionAttemptID attemptID2 = new ExecutionAttemptID();
+
+		final String fileName = "test_file";
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey),
+			true);
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID, attemptID1);
+		fileCache.createTmpFile(fileName, entry, jobID, attemptID2);
+
+		final Path dstPath = copyResult.get();
+		final FileSystem fs = dstPath.getFileSystem();
+		final FileStatus fileStatus = fs.getFileStatus(dstPath);
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID1);
+		// still should be available
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID2);
```

Comment:
are we at risk of race-conditions here between this call and the second `createTmpFile` call? We never wait for the returned future.

Reply:
We are not, cause the `DirectExecutorService` is synchronous for all non-scheduled tasks. The only scheduled task is the `DeleteProcess` though, which we capture and run manually.